cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -286,6 +312,135 @@ private void deployApp(SystemApplication application) throws Exception {
+        continue;
+      }
+
+      String configFilename = action.getConfigFilename();
+      if (configFilename == null) {
+        LOG.error("Config not found for plugin {}", pkg.getName());
+        continue;
+      }
+      URL configUrl = new URL(url, Joiner.on("/").join(Arrays.asList("/v2/packages", pkg.getName(), pkg.getVersion(),
+                                                                     configFilename)));
+      // Download plugin json from hub
+      JsonObject jsonObj = GSON.fromJson(doGetAsString(configUrl), JsonObject.class);
+      List<String> parents = GSON.fromJson(jsonObj.get("parents").toString(), new TypeToken<ArrayList<String>>() {
+      }.getType());
+      String resource = String.format("namespaces/default/artifacts/%s", pkg.getName());
+
+      String jarName = action.getJarName();
+      if (jarName == null) {
+        LOG.error("Jar not found for plugin {}", pkg.getName());
+        continue;
+      }
+      URL jarUrl = new URL(url, Joiner.on("/").join(Arrays.asList("/v2/packages", pkg.getName(), pkg.getVersion(),
+                                                                  jarName)));
+      // Download plugin jar from hub
+      byte[] jar = doGetAsBytes(jarUrl);
+
+      // Install plugin jar
+      HttpRequest.Builder requestBuilder = remoteClient.requestBuilder(HttpMethod.POST, resource)
+        .addHeader("Artifact-Version", pkg.getVersion())
+        .addHeader("Artifact-Extends", Joiner.on("/").join(parents))
+        .addHeader("Content-Type", "application/octet-stream")
+        .withBody(ByteBuffer.wrap(jar));
+      execute(requestBuilder.build());
+
+
+      String properties = GSON.toJson(jsonObj.get("properties"));
+      resource = String.format("namespaces/default/artifacts/%s/versions/%s/properties", pkg.getName(),
+                               pkg.getVersion());
+      // Install plugin properties
+      requestBuilder = remoteClient.requestBuilder(HttpMethod.PUT, resource)
+        .addHeader("Content-Type", "application/json")
+        .withBody(properties);
+      execute(requestBuilder.build());
+    }
+  }
+
+  @VisibleForTesting
+  String doGetAsString(URL url) throws IOException {
+    HttpRequest request = HttpRequest.get(url).build();
+    return execute(request).getResponseBodyAsString();
+  }
+
+  @VisibleForTesting
+  byte[] doGetAsBytes(URL url) throws IOException {
+    HttpRequest request = HttpRequest.get(url).build();
+    return execute(request).getResponseBody();
+  }
+
+  @VisibleForTesting
+  HttpResponse execute(HttpRequest request) throws IOException {
+    return Retries.callWithRetries(() -> {
+      HttpResponse httpResponse = remoteClient.execute(request);
+
+      if (httpResponse.getResponseCode() != HttpURLConnection.HTTP_OK) {
+        if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_CONFLICT) {
+          LOG.debug("Request {}, artifact already exists", request.getURL());
+          return null;
```

Comment:
What does it mean to return null? Generally it is not a good idea.

Reply:
removed. fixed in HttpClient